### DISCUSSION TITLE
Linker error in current dev / libdeps only cloned shallowly?

### DISCUSSION THREAD

**Discussion by mheyse** [emsesp/EMS-ESP32#2039] [General]
Summary: I couldn't build the latest dev branch because `lib_deps` is apparently pulled by a shallow clone which only checks out the first revision of the repo, ignoring all later updates. This leads to a dependency version mismatch (ArduinoJson 7.1.0 vs 7.2.0). Manually updating `lib_deps` fixes the issue (as a workaround).

The details: 

I couldn't build the current dev branch (49050471775b3aebdcfe41d47103276e16421df2). The build failed with a linker error.

```
pio run -e s_16M_P
...
Linking .pio/build/s_16M_P/firmware.elf
/home/mihe/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/s_16M_P/src/web/WebModulesService.cpp.o:(.literal._ZN6emsesp10WebModules4readERS0_N11ArduinoJson8V720PB2210JsonObjectE+0x18): undefined reference to `ModuleLibrary::list(ArduinoJson::V720PB22::JsonObject)'
/home/mihe/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: .pio/build/s_16M_P/src/web/WebModulesService.cpp.o: in function `emsesp::WebModules::read(emsesp::WebModules&, ArduinoJson::V720PB22::JsonObject)':
/home/mihe/EMS-ESP32/src/web/WebModulesService.cpp:55: undefined reference to `ModuleLibrary::list(ArduinoJson::V720PB22::JsonObject)'
collect2: error: ld returned 1 exit status
*** [.pio/build/s_16M_P/firmware.elf] Error 1
```

After digginig into it I found out that there were two versions of ArduinoJson in the project - version 7.2.0 from `lib/ArduinoJson` and version 7.1.0 from `lib_deps`, which is pulled from `https://github.com/emsesp/EMS-ESP-Modules.git` as per `platformio.ini`. Which was strange, because @proddy updated the repo to version 7.2.0 just earlier this week.

Looking at the `EMS-ESP-Modules` directory:

```
~/EMS-ESP32/.pio/libdeps/s_16M_P/EMS-ESP-Modules$ git log
commit cce72904b5b05a775332676c067d8b76c989d984 (grafted, HEAD -> main, origin/main, origin/HEAD)
Author: Proddy <proddy@users.noreply.github.com>
Date:   Tue Jul 23 10:22:45 2024 +0200

    Add files via upload


~/EMS-ESP32/.pio/libdeps/s_16M_P/EMS-ESP-Modules$ nm -C esp32/libModule.a
...
00000000 t _ZN11ArduinoJson8V710PB226detail10StringPool11dereferenceEPKcPNS0_9AllocatorE$isra$15
...
```

So `libModule.a` is referencing ArduinoJson v7.1.0

I could fix this by manualy updating the repo.

```
~/EMS-ESP32/.pio/libdeps/s_16M_P/EMS-ESP-Modules$ git pull
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 8 (delta 1), reused 0 (delta 0), pack-reused 0 (from 0)
Unpacking objects: 100% (8/8), 655.51 KiB | 6.24 MiB/s, done.
From https://github.com/emsesp/EMS-ESP-Modules
   cce7290..caa837e  main       -> origin/main
Updating cce7290..caa837e
Fast-forward
 esp32/libModule.a   | Bin 807282 -> 811086 bytes
 esp32c3/libModule.a | Bin 984740 -> 985240 bytes
 esp32s3/libModule.a | Bin 812690 -> 816554 bytes
 3 files changed, 0 insertions(+), 0 deletions(-)
mihe@LAPTOP-MH:~/EMS-ESP32/.pio/libdeps/s_16M_P/EMS-ESP-Modules$ git log
commit caa837e5b579bd9598801b0a5f5edc49db03ef10 (HEAD -> main, origin/main, origin/HEAD)
Author: Proddy <proddy@users.noreply.github.com>
Date:   Thu Sep 19 10:04:10 2024 +0200

    ArduinoJson 7.2

commit cce72904b5b05a775332676c067d8b76c989d984 (grafted)
Author: Proddy <proddy@users.noreply.github.com>
Date:   Tue Jul 23 10:22:45 2024 +0200

    Add files via upload
 ```
 
 Now the build suceeds.
 
Is this a platform.io issue, maybe intentional? Is a tag or something required to specify that the latest revision of the repo is supposed to be used?

(This is not really related to the issue, but what are you planning to do with the `EMS-ESP-Modules` repo? Do you intend to put everything that's currently in `lib` into this repo, as binaries? Where will the source code be?)

**Comment by proddy** [comment]
thanks for pointing this out. I think what I need to do is update the manifest file at https://github.com/emsesp/EMS-ESP-Modules/blob/main/library.json with a new version like 1.0.1 to trigger the re-fetch

**Reply by proddy** [reply]
think it should be fixed now.

**Reply by mheyse** [reply]
it is. 👍
